auslib/db.py

```diff
@@ -1923,6 +1923,25 @@ def getBlob():
 
         return blob
 
+    def getReleaseBlobs(self, names, transaction=None):
+        blobs = {}
+        uncached = []
+        for name in names:
+            cached_blob = cache.get('blob', name)
+            if cached_blob:
+                blobs.update({name: cached_blob})
+            else:
+                uncached.append(name)
+
+        q = self.select(where=[self.name.in_(tuple(uncached))], columns=[self.name, self.data, self.data_version], transaction=transaction)
+        for row in q:
+# TODO: investigate what's slowing down cache.put
+#            cache.put('blob', row['name'], row['data'])
```

Comment:
Putting items in cache seems to be slowing this down, I suspect it's what was slowing down the `getReleaseBlob` when doing it on more than one blobs. Is there a way of putting a `dict/multiple items` in cache at once instead of how we are currently doing it here https://github.com/mozilla/balrog/blob/5e64240e77d50b6e12c8034728b8c71c3974e7e2/auslib/util/cache.py#L75. Tried look at the documentation of `repoze.lru API` (http://docs.repoze.org/lru/api.html) but couldn't find anything helpful.

Reply:
How slow is slow? And is it only slow on the first request, or slow on subsequent requests as well?

I don't see a way to add multiple items in one call :(.

Reply:
Yeah it should only be slow on first request, but then retrieving an item from `cache` also tends to be slow if the item exists ( this is not the case if item doesn't exist though ). To give you a clear example. When testing on local, 
 - Request takes `~380ms` when not using cache at all
- `~1371ms` on first request when using cache (putting items in cache)
- Then unexpectedly `~1100ms` on next requests when using cache (retrieving items from cache)
The last call made me do some digging and also found out `cache.get(k,v)` is also slow when entry exists. Not yet sure why though.

Reply:
Hmm, ok. Let me think about this

Reply:
Hm, it looks like this was added as footgun protection against a caller putting something into the cache, and then modifying it. I think this is primarily to make sure we don't end up with bad data in the cache when trying to merge blobs together in https://github.com/mozilla/balrog/blob/dd21686a6dbd5b94698f04e518b8d686ad032d95/auslib/db.py#L2012. We could look at disabling the deepcopy for admin if it's truly causing that much performance degradation. Could you try turning it off and see how much of a difference that makes before we go down this road?

Nice detective work!

Reply:
Wow! setting this to `False` significantly improves speed from `~2000ms` to `~600ms` on the first request and from `~1300ms` to `~200ms` on subsequent requests with caching.

Reply:
Wow! I'm seeing similar results here. I'd also be curious to know what the difference is in load times of the Rules with and without this patch -- ultimately what we care about is the load time of that page.

It sounds like this is an angle worth pursuing, though. I think we need to dive a bit deeper into whether or not disabling the deepcopys would cause cache poisoning or not. I'm not sure how easy it is, but enabling the cache and verifying its contents in the tests around that code (https://github.com/mozilla/balrog/blob/master/auslib/test/test_db.py#L1399) would be a good start.